File: src/main/java/org/hisp/dhis/expression/ast/Typed.java
package org.hisp.dhis.expression.ast;

import java.time.LocalDate;
import java.time.ZoneId;
import java.util.Date;

import static java.lang.String.format;

public interface Typed {

    static Double toNumberTypeCoercion(Object value)
    {
        if (value == null) {
            return null;
        }
        if (value instanceof Boolean) {
            return value == Boolean.TRUE ? 1d : 0d;
        }
        if (value instanceof Number) {
            return ((Number) value).doubleValue();
        }
        return Double.valueOf(value.toString());
    }

    static Boolean toBooleanTypeCoercion(Object value)
    {
        if (value == null)
        {
            return null;
        }
        if (value instanceof Boolean) {
            return (Boolean) value;
        }
        if (value instanceof Number) {
            if (!isNonFractionValue((Number) value)) {
                throw new IllegalArgumentException(format("Could not coerce Double '%s' to Boolean", value));
            }
            return ((Number) value).intValue() != 0;
        }
        return Boolean.valueOf(value.toString());
    }

    static LocalDate toDateTypeCoercion(Object value) {
        if (value == null) return null;
        if (value instanceof LocalDate) return (LocalDate) value;
        if (value instanceof String) return LocalDate.parse((String) value);
        if (value instanceof Date) return LocalDate.ofInstant (((Date) value).toInstant(), ZoneId.systemDefault());
        throw new IllegalArgumentException(format("Count not coerce to date: '%s'", value));
    }

    static String toStringTypeCoercion(Object value) {
        return value == null ? null : value.toString();
    }

    static boolean isNonFractionValue(Number value) {
        return value.doubleValue() % 1.0d == 0d;
    }

    ValueType getValueType();
}

File: src/jvmMain/java/org/hisp/dhis/lib/expression/ast/Typed.kt
package org.hisp.dhis.lib.expression.ast

import org.hisp.dhis.lib.expression.spi.ValueType
import org.hisp.dhis.lib.expression.spi.VariableValue
import java.time.LocalDate
import java.time.ZoneId
import java.util.*

interface Typed {

    fun getValueType(): ValueType

    companion object {
        fun toNumberTypeCoercion(value: Any?): Double? {
            if (value == null) return null
            if (value is VariableValue) return toNumberTypeCoercion(value.valueOrDefault())
            return if (value is Boolean) if (value == true) 1.0 else 0.0
            else (value as? Number)?.toDouble() ?: java.lang.Double.valueOf(value.toString())
        }

        fun toBooleanTypeCoercion(value: Any?): Boolean? {
            if (value == null) return null
            if (value is VariableValue) return toBooleanTypeCoercion(value.valueOrDefault())
            if (value is Boolean) return value
            if (value is Number) {
                require(isNonFractionValue(value)) {
                    String.format(
                        "Could not coerce Double '%s' to Boolean",
                        value
                    )
                }
                return value.toInt() != 0
            }
            return java.lang.Boolean.valueOf(value.toString())
        }

        fun toDateTypeCoercion(value: Any?): LocalDate? {
            if (value == null) return null
            if (value is VariableValue) return toDateTypeCoercion(value.valueOrDefault())
            if (value is LocalDate) return value
            if (value is String) return LocalDate.parse(value)
            if (value is Date) return value.toInstant().atZone(ZoneId.systemDefault()).toLocalDate()
            throw IllegalArgumentException(String.format("Count not coerce to date: '%s'", value))
        }

        fun toStringTypeCoercion(value: Any?): String? {
            return if (value == null) null
            else (value as? VariableValue)?.valueOrDefault()?.toString() ?: value.toString()
        }

        fun toMixedTypeTypeCoercion(value: Any?): Any? {
            if (value == null) return null
            return if (value is VariableValue) {
                when (value.valueType()) {
                    ValueType.NUMBER -> toNumberTypeCoercion(
                        value.valueOrDefault()
                    )

                    ValueType.BOOLEAN -> toBooleanTypeCoercion(
                        value.valueOrDefault()
                    )

                    ValueType.DATE -> toDateTypeCoercion(
                        value.valueOrDefault()
                    )

                    else -> toStringTypeCoercion(value.valueOrDefault())
                }
            }
            else value
        }

        fun isNonFractionValue(value: Number): Boolean {
            return value.toDouble() % 1.0 == 0.0
        }
    }
}
